### DISCUSSION TITLE
ideas for helping people work with py5 + PySimpleGUI

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#187] [Ideas]
The simplest, no thread, approach: https://github.com/villares/sketch-a-day/blob/main/2022/sketch_2022_11_10/sketch_2022_11_10.py
(it slows down the animation a bit, Mike warns it can hang too much the CPU)

A first repeating thread approach:

Summary: We can't guarantee that `controls()`, called from the repeating thread, will run before `draw()`. Calling `controls()` in `setup()` doesn't work. So I created a dummy values dict. But I wish people didn't have to do it, to manually initialize values before the first frame-run of `draw()`, because it can get tedious and error prone in larger layouts.

["module mode" version of the code below](https://gist.github.com/villares/c2239dc2ddc1177b6837b5f9d33e4967)
```python
# note this code is writen in py5 "imported mode" style, you'll need to process it through the sketch_runner

import PySimpleGUI as sg

bg = True
values = {'-Text-': '', '-Hue-': 0, '-BG-': False}   # I wish people didn't need this...

def setup():   # py5 calls this once as the sketch starts
    global window
    size(600, 600)
    color_mode(HSB)
    text_size(30)
    layout = [
        [sg.Text('PySimpleGUI FTW!')],
        [sg.Input(key='-Text-', enable_events=True)],
        [sg.Slider(key='-Hue-', range=(0, 255), orientation='h', enable_events=True)],
        [sg.CB('Background', key='-BG-', enable_events=True)]
    ]
    window = sg.Window('Sketch controls!', layout)
    traverse_layout(layout)   # maybe I could initialize values dict here
    launch_repeating_thread(controls, name='controls')  

def controls():
    global event, values
    event, values = window.read()
    if event == sg.WIN_CLOSED:
      exit_sketch()  # triggers py5 exiting cleanup
    
def traverse_layout(layout):  # some exploration I was trying 
    defaults = {
        'slider': 0,
        'checkbox': False,
        }
    for element in layout:
        if isinstance(element, list):
            traverse_layout(element)
        elif element.key:
            print(element.key, defaults.get(element.Type, ''))
            #yield element.key, defaults.get(element.Type, '')

def draw():   # py5 calls this in a loop (it is the "main" py5 drawing/animation loop)
    if values['-BG-']:
        background(50, 100, 100)
    fill(255)
    text(values['-Text-'], 50, 500)
    fill(values['-Hue-'], 200, 200)
    circle(100, 100, 50 + 50 * sin(frame_count / 10))

def exiting():   # called by py5 on exit
    stop_thread('controls')
    window.close()
```

**Comment by villares** [comment]
Dear @PySimpleGUI, is there, or could we have, a way of retrieving an initial `values` dict from a layout, before an actual `window.read()`?

**Comment by PySimpleGUI** [comment]
To get the values dictionary, you'll need to have the window created.  I suggest you set `finalize=True` when making the window, just in case... you may not need it.

Then to get the initial values, call `window.read(timeout=0)`.  This will get you the value dictionary without any delay.

**Reply by PySimpleGUI** [reply]
```python
import PySimpleGUI as sg

layout = [  [sg.Text('My layout')],
            [sg.Button('Go'), sg.Button('Exit')] ]

window = sg.Window('My new window', layout, finalize=True)

# do any operations you want on the window like binding events, drawing, etc, that you want to do prior to the event loop later

# Get the initial values dictionary (I've never seen a case where this was needed, but this how you can do it)

event, values = window.read(timeout=0)

while True:             # Event Loop
    event, values = window.read()
    print(event, values)
    if event in (None, 'Exit'):
        break

window.close()

```

**Comment by villares** [comment]
I see, but this doesn't work for us, because calling
windows.read(timeout=0) outside the thread  crashes. I'll try to add the
error message here. Does it make sense to you why I'd like to have a
pre-initialized values dict before the first frame of the main drawing
loop, right?

**Reply by villares** [reply]
![image](https://user-images.githubusercontent.com/3694604/201474051-7f60ac94-17db-4161-ba7c-dc8fec52320c.png)
```
py5 encountered an error in your code:File "/home/villares/GitHub/sketch-a-day/2022/sketch_2022_11_11/sketch_2022_11_11_crash.py", line 24, in controls
    22   def controls():
    23       global event, values
--> 24       event, values = window.read()
    25       if event == sg.WIN_CLOSED:
    ..................................................
     event = '__TIMEOUT__'
     values = {'-Text-': '',
               '-Hue-': 0.0,
               '-BG-': False}
     window.read = <method 'Window.read' of <PySimpleGUI.PySimpleGUI.Window obj
                    ect at 0x7fe7f83d27a0> PySimpleGUI.py:10039>
    ..................................................

File "/home/villares/.local/lib/python3.10/site-packages/PySimpleGUI/PySimpleGUI.py", line 10075, in read
    10039  def read(self, timeout=None, timeout_key=TIMEOUT_KEY, close=False):
 (...)
    10071  
    10072  
    10073      while True:
    10074          Window._root_running_mainloop = self.TKroot
--> 10075          results = self._read(timeout=timeout, timeout_key=timeout_key)
    10076          if results is not None:
    ..................................................
     self._read = <method 'Window._read' of <PySimpleGUI.PySimpleGUI.Window ob
                   ject at 0x7fe7f83d27a0> PySimpleGUI.py:10105>
     timeout = None
     timeout_key = '__TIMEOUT__'
    ..................................................

File "/home/villares/.local/lib/python3.10/site-packages/PySimpleGUI/PySimpleGUI.py", line 10203, in _read
    10105  def _read(self, timeout=None, timeout_key=TIMEOUT_KEY):
 (...)
    10199          self.CurrentlyRunningMainloop = True
    10200          # self.TKroot.protocol("WM_DESTROY_WINDOW", self._OnClosingCallback)
    10201          # self.TKroot.protocol("WM_DELETE_WINDOW", self._OnClosingCallback)
    10202          Window._window_running_mainloop = self
--> 10203          Window._root_running_mainloop.mainloop()
    10204  
    ..................................................
     Window._root_running_mainloop.mainloop = <method 'Misc.mainloop' of .!toplevel __init__.py:1456>
    ..................................................

File "/home/villares/apps/thonny/lib/python3.10/tkinter/__init__.py", line 1458, in mainloop
    1456  def mainloop(self, n=0):
    1457      """Call the mainloop of Tk."""
--> 1458      self.tk.mainloop(n)
    ..................................................
     n = 0
    ..................................................

---- (full traceback above) ----
File "/home/villares/GitHub/sketch-a-day/2022/sketch_2022_11_11/sketch_2022_11_11_crash.py", line 24, in controls
    event, values = window.read()
File "/home/villares/.local/lib/python3.10/site-packages/PySimpleGUI/PySimpleGUI.py", line 10075, in read
    results = self._read(timeout=timeout, timeout_key=timeout_key)
File "/home/villares/.local/lib/python3.10/site-packages/PySimpleGUI/PySimpleGUI.py", line 10203, in _read
    Window._root_running_mainloop.mainloop()
File "/home/villares/apps/thonny/lib/python3.10/tkinter/__init__.py", line 1458, in mainloop
    self.tk.mainloop(n)

RuntimeError: Calling Tcl from different apartment
```

**Comment by PySimpleGUI** [comment]
I dunno if this is of help, but here's a demo showing getting data through a queue.

No, it does not make sense to me.

```python
import PySimpleGUI as sg
import random
import time
import queue

"""
    Demo - Multi-threaded "Data Pump"

    A thread gets data from a queue object and passes it over to the main event loop 

    Copyright 2022 PySimpleGUI
"""

THREAD_KEY = '-THREAD-'
THREAD_INCOMING_DATA = '-INCOMING DATA-'
THREAD_EXITNG = '-THREAD EXITING-'
THREAD_EXTERNAL_EXITNG = '-EXTERNAL THREAD EXITING-'

thread_queue = queue.Queue()


def external_thread(thread_queue:queue.Queue):
    """
    Represents some external source of data
    :param thread_queue:
    :return:
    """
    i = 0
    while True:
        wait_time = random.randint(0,10)
        time.sleep(wait_time)
        thread_queue.put(f'Incoming data #{i} wait was {wait_time} seconds')
        i += 1

def the_thread(window:sg.Window, thread_queue:queue.Queue):
    """
    The thread that communicates with the application through the window's events.
    Waits for data from a queue and sends that data on to the event loop
    :param window:
    :param thread_queue:
    :return:
    """

    while True:
        data = thread_queue.get()
        window.write_event_value((THREAD_KEY, THREAD_INCOMING_DATA), data)  # Data sent is a tuple of thread name and counter


def main():
    layout = [  [sg.Text('My Simulated Data Pump')],
                [sg.Multiline(size=(80,20), k='-MLINE-')],
                [sg.Button('Go'), sg.Button('Exit')]  ]

    window = sg.Window('Simulated Data Pump', layout, finalize=True, relative_location=(0, -300))
    downloading, max_value = False, 0

    while True:             # Event Loop
        event, values = window.read()
        # print(event, values)
        if event == sg.WIN_CLOSED or event == 'Exit':
            break
        if event == 'Go' and not downloading:
            window.start_thread(lambda: external_thread(thread_queue), (THREAD_KEY, THREAD_EXTERNAL_EXITNG))
            window.start_thread(lambda: the_thread(window, thread_queue), (THREAD_KEY, THREAD_EXITNG))
        # Events coming from the Thread
        elif event[0] == THREAD_KEY:
            if event[1] == THREAD_INCOMING_DATA:
                data = values[event]
                downloading = True
                window['-MLINE-'].print('Incoming data:', c='white on red', end='')
                window['-MLINE-'].print(data)
            elif event[1] == THREAD_EXITNG:
                window['-MLINE-'].print('Thread has exited')
            elif event[1] == THREAD_EXTERNAL_EXITNG:
                window['-MLINE-'].print('Data Pump thread has exited')

    window.close()

if __name__ == '__main__':
    main()

```

**Comment by PySimpleGUI** [comment]
Do you have a simple architecture diagram of your entire application that you have in mind.  That would be likely the best way for me to get up to speed.  And I would also stop guessing at what you might need.  I only knew data was incoming from elsewhere and sent to the event loop.  Not sure what happens in your event loop.

I usually set uninitialized variables to `None` and check for `None` before I use them.

**Reply by villares** [reply]
I see.  I can work with PySimpleGUI and threads as it is now, no problem. My aim here is to find a strategy that makes it easier for my students to use it.

One way, @py5coding suggested, is for the students to check if the GUI has started to conditionally draw anything in the main loop... I thing I might have to settle with it:

```python
import PySimpleGUI as sg

bg = True
values = {}

def setup():   # py5 calls this once as the sketch starts
    global window, event, values 
    size(600, 600)
    color_mode(HSB)
    text_size(30)
    layout = [
        [sg.Text('PySimpleGUI FTW!')],
        [sg.Input(key='-Text-', enable_events=True)],
        [sg.Slider(key='-Hue-', range=(0, 255), orientation='h', enable_events=True)],
        [sg.CB('Background', key='-BG-', default=True, enable_events=True)]
    ]
    window = sg.Window('Sketch controls!', layout)
    launch_repeating_thread(controls, name='controls')  

def controls():
    global event, values
    if frame_count == 0:
        event, values = window.read(timeout=0)
    else:
        event, values = window.read()
    if event == sg.WIN_CLOSED:
      exit_sketch()  # triggers py5 exiting cleanup
    
def draw():   # py5 calls this in a loop (it is the "main" py5 drawing/animation loop)
    if values:
        if values['-BG-']:
            background(50, 100, 100)
        fill(255)
        text(values['-Text-'], 50, 500)
        fill(values['-Hue-'], 200, 200)
        circle(100, 100, 50 + 50 * sin(frame_count / 10))

def exiting():   # called by py5 on exit
    stop_thread('controls')
    window.close()
```
![image](https://user-images.githubusercontent.com/3694604/201474403-8144c650-4388-4c13-b1cf-fc91f55339f7.png)

**Comment by PySimpleGUI** [comment]
If your other event loop is controlling CPU usage by pending, then you can run the event loop in PySimpleGUI with a `timeout=0`... let it immediately return control to whoever is controlling the pace.

**Reply by villares** [reply]
Yes this is the "no thread approach" I linked in the first message , I think...

**Comment by PySimpleGUI** [comment]
```python
import PySimpleGUI as sg
import random
import time
import queue

"""
    Demo - Multi-threaded "Data Pump"

    A thread gets data from a queue object and passes it over to the main event loop 

    Copyright 2022 PySimpleGUI
"""

THREAD_KEY = '-THREAD-'
THREAD_INCOMING_DATA = '-INCOMING DATA-'
THREAD_EXITNG = '-THREAD EXITING-'
THREAD_EXTERNAL_EXITNG = '-EXTERNAL THREAD EXITING-'

thread_queue = queue.Queue()
gsize = (400,400)


def external_thread(thread_queue:queue.Queue):
    """
    Represents some external source of data
    :param thread_queue:
    :return:
    """
    i = 0
    while True:
        time.sleep(.01)
        point = (random.randint(0,gsize[0]), random.randint(0,gsize[1]))
        radius = random.randint(10, 40)
        thread_queue.put((point, radius))
        i += 1

def the_thread(window:sg.Window, thread_queue:queue.Queue):
    """
    The thread that communicates with the application through the window's events.
    Waits for data from a queue and sends that data on to the event loop
    :param window:
    :param thread_queue:
    :return:
    """

    while True:
        data = thread_queue.get()
        window.write_event_value((THREAD_KEY, THREAD_INCOMING_DATA), data)  # Data sent is a tuple of thread name and counter


def main():
    layout = [  [sg.Text('My Simulated Data Pump')],
                [sg.Multiline(size=(80,20), k='-MLINE-')],
                [sg.Graph(gsize, (0, 0), gsize, expand_x=True, expand_y=True, k='-G-', background_color='gray')],

                [sg.Button('Go'), sg.Button('Exit')]  ]
    window = sg.Window('Simulated Data Pump', layout, finalize=True, relative_location=(0, -300))

    graph = window['-G-']       # type: sg.Graph

    while True:             # Event Loop
        event, values = window.read()
        # print(event, values)
        if event == sg.WIN_CLOSED or event == 'Exit':
            break
        if event == 'Go':
            window.start_thread(lambda: external_thread(thread_queue), (THREAD_KEY, THREAD_EXTERNAL_EXITNG))
            window.start_thread(lambda: the_thread(window, thread_queue), (THREAD_KEY, THREAD_EXITNG))
        # Events coming from the Thread
        elif event[0] == THREAD_KEY:
            if event[1] == THREAD_INCOMING_DATA:
                point, radius = values[event]
                graph.draw_circle(point, radius=radius, fill_color='green')
                window['-MLINE-'].print(f'Drawing at {point} radius {radius}', c='white on red')
            elif event[1] == THREAD_EXITNG:
                window['-MLINE-'].print('Thread has exited')
            elif event[1] == THREAD_EXTERNAL_EXITNG:
                window['-MLINE-'].print('Data Pump thread has exited')

    window.close()

if __name__ == '__main__':
    main()

```

Maybe this is closer?  (sorry my screen capture is slow)


How do you want to communicate GUI events back to py5?



![python_IsqbLneRqH](https://user-images.githubusercontent.com/46163555/201474249-96ed6a44-9c7d-4b35-8e3a-8484c7d0ce58.gif)

**Reply by PySimpleGUI** [reply]
HA!  I **knew** you would like an example that drew something
![image](https://user-images.githubusercontent.com/46163555/201474514-c7a7d3a8-72c1-4fba-b5f7-aff5ad63470a.png)

You are allowed to only call `sg.write_event_value` from a thread.  You can't call `window.read` for example.

**Comment by PySimpleGUI** [comment]
The error message is that you're trying to call PySimpleGUI from a thread... no can  do.

**Reply by PySimpleGUI** [reply]
PySimpleGUI **must** be the Python main thread.  tkinter is not capable of being run as a thread.  You'll have to run py5 as a thread.  My examples here show what I mean.... the PySimpleGUI code is running as the main thread.

**Reply by villares** [reply]
Hmmm, this is kind of confusing for me... looks like I'm calling window.read() from the thread in the example I showed here: https://github.com/py5coding/py5generator/discussions/187#discussioncomment-4123377

**Reply by PySimpleGUI** [reply]
Yea, if you are, then you'll have to change it.... normally, like you saw above, tkinter will "catch you" and crash.  Sometimes you can cheat.... until you can't.... never cheat.  It's a pain, but PySimpleGUI has to be the main program.

**Reply by PySimpleGUI** [reply]
```RuntimeError: Calling Tcl from different apartment```

This is one of the "you got caught" errors.

If both programs MUST be the main thread, then you'll have to have each do round-robin scheduling, which kinda sucks, but
![image](https://user-images.githubusercontent.com/46163555/201474760-1d83f1c4-4964-4bd1-8438-5a2a382d59d7.png)
it's one way to get somewhere with it.  Otherwise, you just can't do it.

**Reply by villares** [reply]
Let's see what @py5coding, who has a proper grasp of the framework architecture, can tell us about this discussion!
Thank you very much for your attention!

**Comment by PySimpleGUI** [comment]
> My aim here is to find a strategy that makes it easier for my students to use it.

I would write the data routing code for them and hide it.  Give them a clean interface to work with so they don't see any ugly code.

**Comment by PySimpleGUI** [comment]
I don't know the py5 package, clearly.   I saw in the docs that both PySimpleGUI and py5 are trying to manage some of the same devices, such as the mouse.  If using the Graph element to draw, then the mouse handling is already built-into the Graph element itself.  I'm wondering if there are going to be collisions between tkinter and py5 such that events get lost.

**Comment by hx2A** [comment]
Thank you, @PySimpleGUI , for your assistance and contributions to this discussion.

> Sometimes you can cheat.... until you can't.... never cheat.

Words to live by.

> Do you have a simple architecture diagram of your entire application that you have in mind. That would be likely the best way for me to get up to speed. And I would also stop guessing at what you might need. I only knew data was incoming from elsewhere and sent to the event loop. Not sure what happens in your event loop.

There's a lot I need to explain about py5, threading, and py5's event loop, both here and in the py5 documentation.

First, you should know that py5 is a Python library that is backed by the Processing Java core libraries. It uses [jpype](https://github.com/jpype-project/jpype) to make calls from Python to Java && to make calls from Java back to Python. The later is actually what makes py5 special (and possible).

Here's some pseudocode for how py5's `run_sketch()` method does NOT work:

```python
def run_sketch():
    call_user_setup()
    while True:
        call_user_draw()
```

If this were the case, the thread that makes the call to `run_sketch()` would also be the same thread that is executing `draw()`. 

Here's some pseudocode for how py5's `run_sketch()` method does work:

```python
def run_sketch():
    launch_java_thread_for_running_sketch()
```

The `run_sketch()` method doesn't block and will return control to Python after the window is opened and the animation starts running. In Java's JVM there is an animation thread that is running a loop and calling the user methods. That animation thread is not limited by Python's GIL.

Here's some pseudocode for what is happening on the Java side of py5:

```java
void runAnimation():
    jniCallToUserSetup()
    while (true):
        jniCallToUserDraw()
```

py5's Java code is making JNI calls back to Python to execute the user's `draw()` and `setup()` methods. Those user-defined methods will call py5 methods, many of which will make calls to Java. Those JNI calls to Python are all on the same new thread created by jpype for this purpose.

Try running this code:

```python
import threading
import py5

def setup():
    print('setup', threading.current_thread().getName())

def draw():
    if py5.frame_count < 10:
        print('draw', threading.current_thread().getName())

def mouse_entered():
    print('mouse_entered', threading.current_thread().getName())

print(threading.current_thread().getName())
py5.run_sketch()
```

When run from the IPython REPL, I get:

```txt
MainThread
setup Dummy-1
draw Dummy-1
draw Dummy-1
draw Dummy-1
draw Dummy-1
draw Dummy-1
draw Dummy-1
draw Dummy-1
draw Dummy-1
draw Dummy-1
mouse_entered Dummy-1
mouse_entered Dummy-1
mouse_entered Dummy-1
```

As you can see, all of the user's methods are executed on the same thread that is different from the one I started with.

However, you should know that everything is more complicated on OSX. That operating system requires GUIs to run on the main thread. py5 has to do some extra stuff to run on that platform.

Tragically, when I tried running @villares 's first example code (from the first post) on my Mac, I got a horrific Python seg-fault.

The PySimpleGUI documentation says:

> Effortlessly move across tkinter, Qt, WxPython, and the Web (Remi) by changing only the import statement.

Can we get around the problems discussed in this thread by changing the GUI backend? And do any of them work better on OSX? @villares , I assume you want what you are doing to be cross-platform?

**Reply by hx2A** [reply]
> Here's some pseudocode for how py5's `run_sketch()` method does NOT work:
> 
> ```python
> def run_sketch():
>     call_user_setup()
>     while True:
>         call_user_draw()
> ```

Interestingly, the very first version of py5 did work this way. That only worked for Processing's default JavaAWT renderer though. The OpenGL renderers P2D and P3D can't work that way. OpenGL has this concept of a "context" where only one thread is allowed to do OpenGL stuff. When the user's `draw()` method is making calls back to Java, Java sees it as the same thread. It is because of this that py5 is able to work as well as it does.

**Reply by villares** [reply]
> Can we get around the problems discussed in this thread by changing the GUI backend? And do any of them work better on OSX? @villares , I assume you want what you are doing to be cross-platform?

Sure!  You might want to try `pip install PySimpleGUIQt`... I don't have access to Mac OS easily...

![image](https://user-images.githubusercontent.com/3694604/201536432-2a5786b9-169f-4045-9fe5-de1519882439.png)

Or did you mean the threadless version?

```python
import PySimpleGUIQt as sg      

bg = True

def setup():
    global window
    size(600, 600)
    color_mode(HSB)
    text_size(30)
    layout = [[sg.Text('PySimpleGUI FTW!')],      
          [sg.Input(key='-Text-')],
          [sg.Slider(key='-Hue-', range=(0, 255), orientation='h')], 
          [sg.Button('Background')]]      
    window = sg.Window('Sketch controls!', layout)      

def draw():
    global bg
    event, values = window.read(timeout=5)
    if event == sg.WIN_CLOSED:        
        exit_sketch()
        return
    elif event == 'Background':
        bg = not bg
        background(200)
    
    if bg:
        background(50, 100, 100)
    fill(255)
    text(values['-Text-'], 50, 500)
    fill(values['-Hue-'], 200, 200)
    circle(100, 100, 50 + 50 * sin(frame_count / 10)) 
          
def exiting():
    window.close()
```
